File: LineFind.py
import psycopg2
import time
import random

try:
    conn = psycopg2.connect("dbname='postgres' user='tester' host='localhost' password='hardpass'")
    print ("connect")
except:
    print ("I am unable to connect to the database")

cur = conn.cursor()


def randomList(): #for test (-1 = null, without tag)
	eqip = []
	typeDish = []
	parameter = []

	for i in range(4): eqip.append(random.choice([True,False]))
	for i in range(4): typeDish.append(random.choice([True,False]))
	for i in range(4): parameter.append(random.choice([True,False])) 

	return [random.randint(5,120), #time
	random.randint(-1,20), #kichen
	random.randint(-1,2), #meal
	random.randint(-1,3), #challange
	random.choice([-1,True,False]), #smallCost
	random.choice([-1,True,False]), #exotic
	eqip,
	typeDish,
	parameter]

def lineReqestPars(reqest): #make string sql-command
	s = ""
	if(reqest[0]!=-1):
		s+= "time <= " + str(reqest[0])
	if(reqest[1]!=-1):
		if s: s+= " and "
		s+= "kitchen = " + str(reqest[1])
	if(reqest[2]!=-1):
		if s: s+= " and "
		s+= "(meal >> " + str(reqest[2]) + " & 1) = 1"
	if(reqest[3]!=-1):
		if s: s+= " and "
		s+= "challange = " + str(reqest[3])
	if(reqest[4]!=-1):
		if s: s+= " and "
		s+= "smallCost"
	if(reqest[5]!=-1):
		if s: s+= " and "
		s+= "exotic"

	if(reqest[6][0]):
		if s: s+= " and "
		s+= "eqip[0]"
	if(reqest[6][1]):
		if s: s+= " and "
		s+= "eqip[1]"
	if(reqest[6][2]):
		if s: s+= " and "
		s+= "eqip[2]"
	if(reqest[6][3]):
		if s: s+= " and "
		s+= "eqip[3]"

	if(reqest[7][0]):
		if s: s+= " and "
		s+= "typeDish[0]"
	if(reqest[7][1]):
		if s: s+= " and "
		s+= "typeDish[1]"
	if(reqest[7][2]):
		if s: s+= " and "
		s+= "typeDish[2]"
	if(reqest[7][3]):
		if s: s+= " and "
		s+= "typeDish[3]"

	if(reqest[8][0]):
		if s: s+= " and "
		s+= "parameter[0]"
	if(reqest[8][1]):
		if s: s+= " and "
		s+= "parameter[1]"
	if(reqest[8][2]):
		if s: s+= " and "
		s+= "parameter[2]"
	if(reqest[8][3]):
		if s: s+= " and "
		s+= "parameter[3]"


	return s 

def lineFind(s): #(0.002 - 0.003)
	if s: cur.execute("select name from testbd where " + s + ";")
	else: cur.execute("select name from testbd;")
	

#check test reqest
someReqest = []
for i in range(10):
	print("")
	someReqest.append(randomList()) 
	print(someReqest[i])

	start = time.time()
	s = lineReqestPars(someReqest[i])
	print("%.10f" % (time.time()-start)) #pars time

	start = time.time()
	lineFind(s)
	print("%.10f" % (time.time()-start)) #find time 

	base = cur.fetchall()
	for j in base: print(j)

cur.close()
conn.close()

File: LineFind.go
package main

import (
    "database/sql"
    "fmt"
    //"github.com/lib/pq"
    "bytes"
    "./interact"
    "strconv"
)

func lineFind(s string) []interact.Respond{
	//fmt.Println("reqest = " + s)
	connStr := "user=tester password=hardpass dbname=postgres"
	db, err := sql.Open("postgres", connStr)
	if err != nil{
		fmt.Println(err)
	} else {
		rows, err := db.Query(s);
		if err != nil {
			fmt.Println(err)
		}
		defer db.Close()
		defer rows.Close()

		respond := []interact.Respond{}
		for rows.Next() {
			p := interact.Respond{}
    		err := rows.Scan(&p.Name, &p.Time, &p.Kcal, &p.Kitchen, &p.Challange, &p.Exotic, &p.Smallcost, &p.Eqip, &p.Parameter, &p.TypeDish, &p.Meal, &p.Components, &p.MassComp)
    		if err != nil{
    			fmt.Println(err)
    		} else {
    			respond = append(respond, p)
    		}
		}
		return respond
	}
	return nil
}


func lineParse(req interact.Request) string{

	var l int
	var buf bytes.Buffer

	if req.MaxTime != -1 {
		buf.WriteString("time <= ")
		buf.WriteString(strconv.Itoa(int(req.MaxTime)))
	}

	if req.Kitchen != -1{
		if buf.Len() != 0 {buf.WriteString(" and ")}
		buf.WriteString("kitchen = ")
		buf.WriteString(strconv.Itoa(int(req.Kitchen)))
	}

	if req.Challange != -1{
		if buf.Len() != 0 {buf.WriteString(" and ")}
		buf.WriteString("challange = ")
		buf.WriteString(strconv.Itoa(int(req.Challange)))
	}

	switch req.SmallCost {
	case 1:
		if buf.Len() != 0 {buf.WriteString(" and ")}
		buf.WriteString("smallCost")
	case 0:
		if buf.Len() != 0 {buf.WriteString(" and ")}
		buf.WriteString("not smallCost")
	}

	switch req.Exotic {
	case 1:
		if buf.Len() != 0 {buf.WriteString(" and ")}
		buf.WriteString("exotic")
	case 0:
		if buf.Len() != 0 {buf.WriteString(" and ")}
		buf.WriteString("not exotic")
	}

	if req.Meal != -1{
		if buf.Len() != 0 {buf.WriteString(" and ")}
		buf.WriteString("(meal >> ")
		buf.WriteString(strconv.Itoa(int(req.Meal)))
		buf.WriteString(" & 1) = 1")
	}

	if req.Eqip[0] {
		l = len(req.Eqip)
		for i := 1; i < l; i++{
			if !req.Eqip[i]{
				if buf.Len() != 0 {buf.WriteString(" and ")}
				buf.WriteString("not eqip[")
				buf.WriteString(strconv.Itoa(i-1))
				buf.WriteString("]")
			}
		}
	}

	l = len(req.TypeDish)
	for i := 0; i < l; i++{
		switch req.TypeDish[i] {
		case 1:
			if buf.Len() != 0 {buf.WriteString(" and ")}
			buf.WriteString("typeDish[")
			buf.WriteString(strconv.Itoa(i))
			buf.WriteString("]")
		case 0:
			if buf.Len() != 0 {buf.WriteString(" and ")}
			buf.WriteString("not typeDish[")
			buf.WriteString(strconv.Itoa(i))
			buf.WriteString("]")
		}
	}

	l = len(req.Parameter)
	for i := 0; i < l; i++{
		switch req.Parameter[i] {
		case 1:
			if buf.Len() != 0 {buf.WriteString(" and ")}
			buf.WriteString("Parameter[")
			buf.WriteString(strconv.Itoa(i))
			buf.WriteString("]")
		case 0:
			if buf.Len() != 0 {buf.WriteString(" and ")}
			buf.WriteString("not Parameter[")
			buf.WriteString(strconv.Itoa(i))
			buf.WriteString("]")
		}
	}

	if buf.Len() != 0 { return "select * from testbd where " + buf.String() + ";" }
	return "select * from testbd;"
}

func main(){
	fmt.Println("start")
	
	//t := interact.GenTest(5)
	//for i:=0; i< len(t); i++ {fmt.Println(lineParse1(t[i]))}

	interact.TimeTest(lineFind,lineParse,interact.GenTest(100))

	
	
}

